rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,258 @@
+* Semantically meaningful (to translate it to visual charts and dashboards.)
+
+### Initial Clients
+
+The initial proposal is to provide observability via two clients:
+
+1. `vector top` / `vector tap` CLI commands.
+2. A real-time web UI
+
+This RFC focuses on the web UI, but applies equally to the CLI client due to the possibility of observing a remote Vector instance.
+
+### Protocol
+
+This section summarizes various communications protocols, and their disadvantages for Vector observability:
+
+**REST:**
+
+* It's not suited for streaming data.
+* It requires polling and/or long-lived connections.
+* Data will initially be read-only; many HTTP verbs don't apply to us.
+* The API interface isn't typed by default, making it cumbersome to work on the front-end without additional tooling.
+
+**gRPC:**
+
+* The full gRPC spec is incompatible with current browsers (see [this article](https://grpc.io/blog/state-of-grpc-web/) for a useful summary.)
+* gRPC-Web offers a subset of gRPC features, including a lack of two-way messaging.
+* The TLS story is harder to solve.
+* gRPC-Web support is currently lacking in Rust. There's an [open issue on tower-grpc](https://github.com/tower-rs/tower-grpc/issues/35) to track.
+* Enabling gRPC-Web therefore presently requires a middle-tier proxy, such as [Envoy](https://www.envoyproxy.io/), to transform requests to the gRPC-Web spec.
+
+**WebSockets:**
+
+* Asynchronous data payloads that are suited to streaming data, but harder to reconcile for single requests.
+* Related: The need to match up a 'request ID' with a 'response ID', to determine which payload belongs to a given request.
+* Long-lived connections that require explicit (re)connection logic.
+* Not intrinsically type-safe for the client to consume; high degree of runtime validation.
+* Non-trivial for non-web clients to interact with; not a common protocol outside of the web.
+
+## Internal Proposal
+
+I propose using GraphQL for API communications.
+
+Advantages:
+
+* A known spec with 5 years of history.
+* Type safe.
+* Dual HTTP + WebSockets model suits one-time and streaming queries.
+* Production use at companies with heavy API workloads. Article refs: [Github](https://github.blog/2016-09-14-the-github-graphql-api/), [Facebook](https://www.apollographql.com/blog/graphql-at-facebook-by-dan-schafer-38d65ef075af/), [Shopify](https://shopify.dev/concepts/graphql), [Intuit](https://medium.com/intuit-engineering/graphql-intuits-path-to-one-api-system-b8495e4dd281), [Airbnb](https://medium.com/airbnb-engineering/how-airbnb-is-moving-10x-faster-at-scale-with-graphql-and-apollo-aa4ec92d69e2), [Trello / Atlassian](https://www.atlassian.com/engineering/a-look-at-trello-adopting-graphql-and-apollo-in-a-legacy-application).
+* Rich ecosystem of tooling, for server/clients.
+* Trivial debugging story.
+* No proxy/middle-tier required.
+* Works over plain HTTP.
+* Optional TLS.
```

Comment:
What potential auth is available?

Reply:
We can use whatever is available to us in the intersection of HTTP + WebSockets.

With browsers locking down headers on WS `Upgrade` requests, in Alloy we passed a JWT in the initial connection, which was then scoped to the life of the subscription.

A `query` is just a HTTP POST request, so we'd typically pass that in an `Authorization: Bearer <jwt>` header.

In both cases, we validated the JWT, looked up the session in the DB, and inserted the session in request context so it could be referenced by GraphQL resolvers.

I played with this a little over the weekend using a sqlx backend, and async-graphql's [Context](https://async-graphql.github.io/async-graphql/en/context.html) to feed a 'user' record through context, and it worked out to be pretty terse. I didn't attempt WebSocket authentication yet, but I anticipate it following the same flow.

Reply:
Sorry - I should have been clearer. :) I think we should put something in the RFC covering auth - the above ^^^ is pretty close.

Reply:
Will do 👍

Reply:
Added in 927e7d06d03e905dae4d5fdbd91a90f05aa2fa31